mysql-test/suite/rpl/t/rpl_bug85158.test

```diff
@@ -0,0 +1,108 @@
+#
+# Testcase for bug 85158.
+#
+--source include/not_group_replication_plugin.inc
+--source include/have_gtid.inc
+--source include/have_binlog_format_row.inc
+
+--let $rpl_multi_source= 1
+--let $rpl_topology= 1->2,2->3,1->3
+--source include/rpl_init.inc
+
+# mi writes threshold, if the difference between count of writes to
+# mi repository before and after the actions is greater than the
+# threshold, the test failed
+--let $threshold=5
+
+# The number of inserts on server_1, if the bug is not fixed the amount
+# of mi repository writes will be greater or equal to this value
+--let $inserts_number=50
+
+--echo ###
+--echo # STOP SLAVE ON 2
+--echo ######
+--let $rpl_connection_name= server_2
+--source include/rpl_connection.inc
+STOP SLAVE;
+
+--echo ###
+--echo # CREATE TABLE AND INSERTS ON 1
+--echo ######
+--let $rpl_connection_name= server_1
+--source include/rpl_connection.inc
+CREATE TABLE t1(f1 INT, f2 BLOB);
+--disable_query_log
+--let $j=$inserts_number
+while($j)
+{
+--eval INSERT INTO t1(f1) VALUES($j)
+--dec $j
+}
+--enable_query_log
+
+--echo ###
+--echo # SYNC 1->3
+--echo ######
+--let $sync_slave_connection= server_3
+--source include/sync_slave_sql_with_master.inc
+
+--echo ###
+--echo # MEASURE 3
+--echo ######
+STOP SLAVE;
+UPDATE performance_schema.setup_objects
+  SET enabled='yes', timed='yes'
+  WHERE object_schema='mysql';
+START SLAVE;
+
+let $before=
+`SELECT count_write
+  FROM performance_schema.table_io_waits_summary_by_table
+  WHERE object_schema='mysql' AND object_name ='slave_master_info'`;
+
+--echo ###
+--echo # START SLAVES 2
+--echo ######
+--let $rpl_connection_name= server_2
+--source include/rpl_connection.inc
+START SLAVE;
+
+--echo ###
+--echo # SYNC 1->2,2->3
+--echo ######
+--let $rpl_connection_name= server_1
+--source include/rpl_connection.inc
+--let $sync_slave_connection= server_2
+--source include/sync_slave_sql_with_master.inc
+--let $sync_slave_connection= server_3
+--source include/sync_slave_sql_with_master.inc
+
+--echo ###
+--echo # MEASURE 3
+--echo ######
+let $after=
+`SELECT count_write
+  FROM performance_schema.table_io_waits_summary_by_table
+  WHERE object_schema='mysql' AND object_name ='slave_master_info'`;
+
+--echo ###
+--echo # COMPARE THE MEASUMENT RESULTS
+--echo ######
+--disable_query_log
+--eval SET @result = (($after - $before) < $threshold)
+SELECT @result;
+--enable_query_log
+
+--echo ###
+--echo # CLEANUP
+--echo ######
+UPDATE performance_schema.setup_objects
+  SET enabled='no', timed='no'
+  WHERE object_schema='mysql';
+--let $rpl_connection_name= server_1
+--source include/rpl_connection.inc
+DROP TABLE t1;
+--disable_query_log
+--let $rpl_skip_sync= 1
```

Comment:
Why? Doesn't that leave t1 at servers 2 and 3?

Reply:
I can't prove it right now because I need some time to research the code, but I was pretty sure that for servers with custom configs data directory is wiped out after some test execution.

Reply:
Look at the following stack trace for mysql-test-run.pl:

```
server_need_restart()
servers_need_restart()
run_testcase()
```

Look at the run_testcase():

```
sub run_testcase ($) {
...
  if ( ! using_extern() )
  {
    my @restart= servers_need_restart($tinfo);
    if ( @restart != 0) {
      stop_servers($tinfo, @restart );
    }

    if ( started(all_servers()) == 0 )
    {

      # Remove old datadirs
      clean_datadir() unless $opt_start_dirty;
    ...
    }
...
}
```

And then look at the server_need_restart():
```
sub server_need_restart {
...
  if ( $tinfo->{template_path} ne $current_config_name)
  {
    mtr_verbose_restart($server, "using different config file");
    return 1;
  }
...
}
```

So there is no need to restore server state if it has custom config.

Reply:
I don't understand: what is the relation between rpl_sync setting for rpl_end.inc and server restarts?

Reply:
This question still stands